packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,67 +165,213 @@ function makeSwingStore(dirPath, forceReset = false) {
+          );
+          closeStream(streamName);
+          assert(itemCount === 0, X`leftover item count ${q(itemCount)}`);
         }
-        fs.closeSync(fd);
-      }
-    } catch (e) {
-      if (e.code !== 'ENOENT') {
-        throw e;
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @param {string} streamName  The stream to be written
+   *
+   * @returns {StreamWriter} a writer for the named stream
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      status === 'unused' || !status,
+      X`can't write stream ${q(streamName)} because it's already in use`,
     );
-    assert(streamName.match(/^[-\w]+$/));
-    let fd = streamFds.get(streamName);
-    if (!fd) {
-      fd = fs.openSync(`${dirPath}/${streamName}`, 'a');
-      streamFds.set(streamName, fd);
+    streamStatus.set(streamName, 'write');
+
+    // XXX fdTemp is a workaround for a flaw in TypeScript's type inference
+    // It should be fd, which it should be changed to when and if they fix tsc.
+    let fdTemp = streamFds.get(streamName);
+    if (!fdTemp) {
+      const filePath = `${dirPath}/${streamName}`;
+      const mode = fs.existsSync(filePath) ? 'r+' : 'w';
+      fdTemp = fs.openSync(filePath, mode);
+      streamFds.set(streamName, fdTemp);
     }
+    const fd = fdTemp;
     activeStreamFds.add(fd);
-    fs.writeSync(fd, item);
-    fs.writeSync(fd, '\n');
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
+     *
+     * @returns {Object} the new position after writing
+     */
+    function write(item, position) {
+      assert.typeof(item, 'string');
+      assert(
+        streamFds.get(streamName) === fd,
+        X`can't write to closed stream ${q(streamName)}`,
+      );
+      if (!position) {
+        position = STREAM_START;
+      }
+      const buf = encoder.encode(`${item}\n`);
+      fs.writeSync(fd, buf, 0, buf.length, position.offset);
+      fs.fsyncSync(fd);
+      return {
+        offset: position.offset + buf.length,
+        itemCount: position.itemCount + 1,
+      };
+    }
+    return write;
   }
 
-  const streamStore = { appendItem, readStream };
+  const streamStore = {
+    openReadStream,
+    openWriteStream,
+    closeStream,
+    STREAM_START,
+  };
 
   /**
    * Commit unsaved changes.
    */
   function commit() {
     if (txn) {
       for (const fd of activeStreamFds) {
-        fs.fsyncSync(fd);
+        closefd(fd);
```

Comment:
why are the stream changes inside the `if (txn)` clause? if someone calls `commit()` without being in a txn, should the stream files be left open?

Reply:
Hmm.  Actually, upon reflection I'm wondering if `commit` should be closing the files at all.  The purpose of `commit` is to say "the state we have now is the official state", not to close the db, which means we should be syncing but not closing.  The expectation is that after a transaction is committed there will be another transaction, which will manifest in the stream store as further writes to the file.  And since our means of ensuring a consistent end-of-stream is to write the position to the adjacent transactional database, the important thing for streams at the commit point is just to be sure that any bits that should be on disk actually are, not that the overall file dimensions are correct.  The important thing is that abort can't roll back stream changes, but it will roll back our belief about what the end of stream is which has the same effect.

On the other hand, if we interpret `commit` to mean "we are done with this for a while" (e.g., later cranks may well be for other vats, hence different streams), then closing on commit unconditionally makes sense, since what we are doing is essentially allowing reclamation of the file descriptor.  If the very next thing we do is write to the same stream, the file will just be reopened.

This suggests that in the future, if we want to speed things up, we might want to be doing come kind of file descriptor caching thing, and `commit` would be the point at which the implementation marks the fd as not being the one in active use (and thus eligible to become a member of the cache instead of *the* fd).

For now I think closing unconditionally may be the way to go, though we'll want to watch to see if there's a performance hit.

Reply:
Further thought: I think that in the pre-PR code the `if (txn) {...` conditional was to make sure we didn't do operations on the transaction object when there was no transaction object. This also argues for moving the file stuff outside the if.

Reply:
Final thoughts on this thread: I've changed to sync on commit and close on close.  It will now keep open fds for all active write streams.  This makes everything simpler and is probably the fastest thing anyway.  You can release the fd by closing the stream, which I think will go with swapping out the corresponding vat.  If the issue is too many fds open, we probably have more live vats than we can fit in memory and we're already in trouble.

Reply:
that sounds good: one fd per online vat

Sync needs to happen just before commit, but yeah we can leave the fds open as long as that doesn't interfere with the sync. We'll be commiting once per block, but touching a random collection of vats in between commit points, probably with some sort of power-law distribution (lots of writes for a small number of vats, plus some long tail).